Implement some gz functions

rzinsly opened the pull request:
This PR adds NX versions for gzopen, gzclose, gzwrite and gzread.
I'm hitting CC 66 with `nx_gzread` both on libnxz and zlib compressed files:
```
pid 2855471: Info: uncompress begin: len 16384
pid 2855471: nx_inflate.c:1367: Error: error: cc = 66, cc = 0x42
pid 2855471: nx_inflate.c:1370: Error: CSB: 0x 80 00 42 60
pid 2855471: nx_inflate.c:1370: Error: CSB: 0x 80 00 42 60
pid 2855471: nx_inflate.c:1370: Error: CSB: 0x 80 00 42 60
pid 2855471: nx_inflate.c:1370: Error: CSB: 0x 80 00 42 60
pid 2855471: nx_inflate.c:1690: Error: rc -3
```
I'll also add tests for these new functions.

tuliom commented on lib/nx_gzlib.c:
```diff
@@ -0,0 +1,105 @@
+	strm = &(state->strm);
+	memset(strm, 0, sizeof(z_stream));
+	if(strchr(mode, 'w')){
+		err = nx_deflateInit2_(&(state->strm), Z_DEFAULT_COMPRESSION, Z_DEFLATED, 31, DEF_MEM_LEVEL, Z_DEFAULT_STRATEGY, ZLIB_VERSION, (int)sizeof(z_stream));
```
This is a very long line.
Could you limit it to 80 characters, please?

mscastanho replied:
@rzinsly I added `nx_deflateInit2` as part of an upcoming PR, but I split the commit into #139 in case you want to use it here. It will help make this line shorter.

tuliom commented on lib/nx_gzlib.c:
```diff
@@ -0,0 +1,105 @@
+
+	ret = nx_inflateEnd(strm);
+
+	return ret;
```
I expected to see a `free(file)` somewhere here.

rzinsly replied:
Yes, I'll fix that for the next version.

tuliom commented on lib/nx_gzlib.c:
```diff
@@ -103,3 +103,24 @@ gzFile nx_gzopen_(const char* path, int fd, const char *mode)
+	z_streamp stream = &(state->strm);
+	int rc;
+	uLong last_total_in = stream->total_in;
+	unsigned char* next_out = malloc(sizeof(char)*len);
```
Where is this being freed?

rzinsly replied:
It isn't, I'll fix that for the next version.

tuliom commented on lib/nx_gzlib.c:
```diff
@@ -104,6 +104,29 @@ gzFile nx_gzopen_(const char* path, int fd, const char *mode)
+	z_streamp stream = &(state->strm);
+	int ret, err;
+	uLong last_total_out = stream->total_out;
+	unsigned char* next_in = malloc(sizeof(char)*len);
```
Where is this being freed?

rzinsly commented:
I have a functional version of the patches now, however there is two potential issues:

1. We don't support different strategies now, so if someone set them on `gzopen` `deflateInit` will fail, perhaps we should not set them yet.
2. We don't know how much source will be needed on `gzread` to decompress `len` bytes, reading more from the file will make `gzread` faster however if it's too much we will overflow `fifo_in`. I'm using 10 because it's safe as even sparse1000M doesn't overflow that and it is unlikely there will be a file that compress more than sparse1000M. Reading one byte per loop and inflating with `Z_SYNC_FLUSH` is the safest option but it is slow.

tuliom commented on test/libnxz.abi:
```diff
@@ -24,6 +24,11 @@
+    <elf-symbol name='gzdopen' type='func-type' binding='global-binding' visibility='default-visibility' is-defined='yes'/>
+    <elf-symbol name='gzopen' type='func-type' binding='global-binding' visibility='default-visibility' is-defined='yes'/>
+    <elf-symbol name='gzread' type='func-type' binding='global-binding' visibility='default-visibility' is-defined='yes'/>
+    <elf-symbol name='gzwrite' type='func-type' binding='global-binding' visibility='default-visibility' is-defined='yes'/>
```
I'm afraid this patch has to be split and its content to be distributed across the previous patches in order to guarantee the previous patches do not break `git-bisect`.

rzinsly commented on lib/nx_gzlib.c:
```diff
@@ -155,6 +156,44 @@ gzFile nx_gzopen_(const char* path, int fd, const char *mode)
+	if (stream->avail_in > 0){
+		nx_streamp s = (nx_streamp) stream->state;
+		copy_data_to_fifo_in(s);
+	}
```
Because `gzread` returns when `len` bytes are wrote on `buf`, if we still have unprocessed data on `avail_in` we need to save it or we will lose it, IMHO `fifo_in` seems the best place to save it. Another option is to skip `read` when `avail_in` is not 0, but we can't guarantee `avail_in` will not be overwritten somewhere else between `gzread` calls.

mscastanho replied:
@rzinsly another option is to `lseek` the file offset `avail_in` bytes back, so the next `read` will get the unprocessed bytes. The downside is that we would be re-reading these bytes from disk in the future, while a copy in memory would most likely be cheaper.

tuliom replied:
I like what @rzinsly did here, but I think the code deserves a source code comment that is a copy if his explanation here.

rzinsly replied:
I added a source code comment, `lseek` is not an option because `gzread` needs to be able to read from `stdin`.

rzinsly commented on lib/nx_gzlib.c:
```diff
@@ -0,0 +1,156 @@
+{
+	gz_statep state = (gz_statep) file;
+	z_streamp stream = &(state->strm);
+	unsigned char* next_out = malloc(sizeof(char)*MAX_LEN);
```
Oops, fixed!

mscastanho replied:
Should check if `next_out == NULL`.

mscastanho commented on lib/nx_gzlib.c:
```diff
@@ -0,0 +1,295 @@
+
+			nx_deflate(stream, Z_FINISH);
+			write(state->fd, next_out, MAX_LEN - stream->avail_out);
+		} while (stream->avail_out == 0);
```
If the pending data is exactly `MAX_LEN`, this will execute an extra unneeded loop iteration.

It would be better to change this check to `while (ret != Z_STREAM_END)`, where `ret` would be the return value from `nx_deflate`.

rzinsly replied:
Ok, it's unlikely to happen but it can, I'll change that.

mscastanho commented on lib/nx_inflate.c:
```diff
@@ -1692,13 +1658,18 @@ int nx_inflateSetDictionary(z_streamp strm, const Bytef *dictionary, uInt dictLe
 		return Z_STREAM_ERROR;
 
+	if (s->wrap == (HEADER_ZLIB | HEADER_GZIP)) {
+		prt_err("inflateSetDictionary error: inflate did not dected the header yet.\n");
```
I suspect by `dected` you meant `detect`?

mscastanho commented on lib/nx_zlib.h:
```diff
@@ -113,11 +113,12 @@ struct selector {
 #endif
 #ifndef DEF_WBITS
-#  define DEF_WBITS MAX_WBITS
+#  define DEF_WBITS 47
```
Is this change necessary? The new value of `MAX_WBITS` would have set `47` here as well.

rzinsly replied:
Seems zlib is defining `MAX_WBITS` somewhere. Without this change, at least in the servers I tested, `DEF_WBITS` is always 15 and libnxz assumes we have a zlib header.

mscastanho commented on lib/nx_gzlib.c:
```diff
@@ -0,0 +1,295 @@
+	void* stream;
+
+	state = malloc(sizeof(struct gz_state));
+	memset(state, 0, sizeof(struct gz_state));
```
This line is being added in the 4th commit. Doesn't it belong with the first one that adds `nx_gzopen`?

mscastanho commented on lib/nx_gzlib.c:
```diff
@@ -0,0 +1,295 @@
+	gz_statep state = (gz_statep) file;
+	z_streamp stream = &(state->strm);
+	int ret, err;
+	const uInt max = len > 10? 10 : 1;
```
Is this correct? Why are you using such low values?

rzinsly replied:
Yes, that's one of the points I raised on https://github.com/libnxz/power-gzip/pull/138#issuecomment-1077786216

mscastanho reviewed (changes requested):
Haven't looked at the test yet, but some early comments below.

rzinsly commented on lib/nx_gzlib.c:
```diff
@@ -0,0 +1,212 @@
+gzFile nx_gzopen_(const char* path, int fd, const char *mode);
+
+typedef struct gz_state {
+	gzFile gz;
```
I'm not sure I understand your question.
`gzFile` is a pointer to the semi-opaque structure `gzFile_s`, and it is available on `zlib.h`. What is not available is `gz_state`, that is the full file state data structure that we are defining here.

mscastanho replied:
> gzFile is a pointer to the semi-opaque structure gzFile_s

You're right, I missed this part. Thanks for clarifying.

tuliom commented:
>     2. We don't know how much source will be needed on `gzread` to decompress `len` bytes, reading more from the file will make `gzread` faster however if it's too much we will overflow `fifo_in`. I'm using 10 because it's safe as even sparse1000M doesn't overflow that and it is unlikely there will be a file that compress more than sparse1000M. Reading one byte per loop and inflating with `Z_SYNC_FLUSH` is the safest option but it is slow.

@rzinsly `nx_inflate_()` shouldn't be overflowing too much data to `fifo_in`.  That's part of the issue in #151 . I plan to fix this with PR #150.  With that said: I think it's safe for you to read as much input as possible.

mscastanho commented on test/test_gz.c:
```diff
@@ -0,0 +1,189 @@
+#define DATALEN 1024*1024 // 1MB
+#define BUFLEN	8192
+
+//const char filename[14] = "test_gzfile.gz";
```
Delete this comment

mscastanho commented on test/Makefile.am:
```diff
@@ -86,3 +88,6 @@ regen-abi:
 build-TESTS: $(check_PROGRAMS) $(check_SCRIPTS)
+
+clean-local:
+	-rm -f test_gzfile*.gz
```
This rule is not needed since the test is already deleting the `test_gzfile*.gz` files.

rzinsly replied:
It's just to guarantee make clean cleans everything, if the test do not finish the files will stay there.

mscastanho commented on test/test_gz.c:
```diff
@@ -0,0 +1,189 @@
+	gzFile file;
+	Byte *cur = src;
+	int rc;
+	char buf[BUFLEN];
```
Why is this buffer used to read the data from `src` here? Can't we pass `src` directly to `gzwrite` like in `test_nx_gzwrite`?

rzinsly replied:
`test_nx_gzwrite` writes all data in one `gzwrite` call, buf if we pass `src` to `gzwrite` here each loop iteration will overwrite the data from the previous iteration.

rzinsly replied:
Anyway, we don't really need `buf` as we can pass `cur` to `gzwrite`.
I'll change that, thanks!

mscastanho reviewed (changes requested):
The test fails to compile with `-Wall -Werror`:
```
test_gz.c: In function ‘main’:
test_gz.c:143:6: error: unused variable ‘pid’ [-Werror=unused-variable]
  143 |  int pid;
      |      ^~~
test_gz.c:141:21: error: unused variable ‘compr_len’ [-Werror=unused-variable]
  141 |  const unsigned int compr_len = DATALEN*2;
      |                     ^~~~~~~~~
```

rzinsly commented on test/test_gz.c:
```diff
@@ -0,0 +1,189 @@
+	gzFile file;
+	Byte *cur = dest;
+	int rc;
+	char buf[BUFLEN];
```
Fixed!

mscastanho commented on lib/nx_gzlib.c:
```diff
@@ -0,0 +1,299 @@
+	int ret, err;
+	const uInt max = len > 10? 10 : 1;
+	uLong last_total_out = stream->total_out;
+	unsigned char* next_in = malloc(sizeof(char)*max);
```
Should check if `next_in == NULL`

mscastanho commented on lib/nx_gzlib.c:
```diff
@@ -0,0 +1,299 @@
+	z_streamp stream = &(state->strm);
+	int rc;
+	uLong last_total_in = stream->total_in;
+	unsigned char* next_out = malloc(sizeof(char)*len);
```
Should check if `next_out == NULL`

mscastanho reviewed (changes requested):
Just some minor suggestions. After that, LGTM.

tuliom commented on lib/nx_gzlib.c:
```diff
@@ -162,6 +162,49 @@ gzFile nx_gzopen_(const char* path, int fd, const char *mode)
+
+		if (err != Z_OK && err != Z_STREAM_END){
+			state->err = err;
+			return 0;
```
Returning here will cause a memory leak for `next_in`.

rzinsly replied:
Fixed by freeing it there.

tuliom reviewed (changes requested):
@rzinsly I'm afraid there are still 3 leaks here. Otherwise, LGTM.

rzinsly commented on lib/nx_gzlib.c:
```diff
@@ -0,0 +1,163 @@
+	if (file == NULL){
+		errno = EINVAL;
+		return Z_STREAM_ERROR;
+	}
```
Done!

rzinsly commented on lib/nx_gzlib.c:
```diff
@@ -0,0 +1,163 @@
+		else
+			state->fd = -1;
+	} else {
+		FILE* fp = fopen(path, mode);
```
Done!

rzinsly commented:
I rebased the code on top of current develop and was able to change the input size of `gzread` to half `len` making it way faster.
I also removed the `deflate` boolean from `gz_state` in the third commit to be able to set it when we use the sw `gzopen`.
